accepted/0000-overrides.md

```diff
@@ -0,0 +1,647 @@
+    "webpack@2.x": "2.7.0"
+  }
+}
+```
+
+In this case, we are saying that any `webpack@1` dependencies should be
+bumped up to `webpack@2`, and furthermore, that any `webpack@2`
+dependencies should be pinned to `webpack@2.7.0`.
+
+In practice, since rules are applied once and not stacked or cascaded, any
+webpack dependency that would resolve to a version matched by `2.x` will be
+overridden to `2.7.0`.  But, any dependency that resolves to a version
+matched by `1.x` will be set to whichever version happens to be the latest
+`2.x` at the time of installation.
+
+In order to produce the intended behavior described, the user would have to
+either specify it twice:
+
+```json
+{
+  "overrides": {
+    "webpack@1.x": "2.7.0",
+    "webpack@2.x": "2.7.0"
+  }
+}
+```
+
+Or make the dependency matching range wider:
+
+```json
+{
+  "overrides": {
+    "webpack@1.x || 2.x": "2.7.0"
+  }
+}
+```
+
+### Object Overrides
+
+An object value in an overrides object defines a _child rule set_.
+
+If the first match for a given resolution is an object, then the object is
+a new rule set applied to all resolutions down that path in the dependency
+graph.
+
+For example, this override rule will set all versions of `bar`, but only
+those depended upon by `foo`.
+
+```json
+{
+  "foo": {
+    "bar": "1.2.3"
+  }
+}
+```
+
+Like string overrides, object overrides are _only_ applied if they are the
+first rule in the set to match a given package.
+
+```js
+{
+  "foo": "1.2.3",
+  "foo@1.2.3": {
+    "bar": "2.3.4" // <-- this is never applied anywhere!
+  }
+}
+```
+
+In this case, because the `foo` rule will always match before the
+`foo@1.2.3` rule, and take precedence.
+
+### `"."` Key Within Object Overrides
+
+In order to both the package being targeted and its dependents, the special
+key `"."` can be used within an object override rule set.  For example, to
+set all versions of `foo` to `1.2.3` _and also_ set `bar` to `2.3.4` when
+depended upon by `foo`, this ruleset could be used:
+
+```json
+{
+  "foo": {
+    ".": "1.2.3",
+    "bar": "2.3.4"
+  }
+}
+```
+
+The `"."` key is not relevant in the root overrides rule set, as the root
+package is not ever subject to dependency resolution.
+
+Thus, these two override rulesets are equivalent:
+
+```js
+// string-valued rule
+{
+  "foo": "1.0.0"
+}
+// object with . member
+{
+  "foo": {
+    ".": "1.0.0"
+  }
+}
+```
+
+The `"."` member _must_ have a string value, to prevent ambiguous
+resolutions.
+
+```js
+// ambiguous and invalid!
+{
+  "foo": {
+    ".": { // <-- raises error, "." must be a string value
+      "bar": "1.0.0"
+    },
+    "bar": "2.0.0"
+  }
+}
+```
+
+### Rules in Nested Rule Sets Override Parent Rules
+
+Parent rules are inherited by nested rule sets, applied _after_ the child
+rules.
+
+For example:
+
+```json
+{
+  "foo": {
+    ".": "1.0.0",
+    "bar": {
+      ".": "2.3.4",
+      "baz": "3.0.0"
+    },
+    "baz": "2.0.0",
+    "boo": "3.0.0"
+  },
+  "boo": "1.0.0"
+}
```

Comment:
Yarn also allows this complex overrides but I can't imagine a real world example when you'd want this.

Also, I don't think we could support something like the in pnpm.

We could support only overrides when the direct dependent is specified.

Reply:
Ah, I see, because you might have two instances of the same version of a package subject to different overrides?  Eg:

```
root -> (a, b)
a -> (x@1.2.3)
b -> (x@1.2.3)
x -> (y)

overrides: {
  a: {
    x: {
      y: "1.2.3"
    }
  },
  b: {
    x: {
      y: "2.3.4"
    }
  }
}
```

So, the `x@1.2.3` depended upon by `a` would need a `y` of `1.2.3`, but the `x@1.2.3` depended upon by `b` would need `y@2.3.4`.

Meaning both `root -> a -> x` and `root -> b -> x` could not be symbolic links to the same physical `x@1.2.3` folder.  The realpath target would have to be unique to both the package resolution and its set of overrides.

Am I understanding that correctly?  If so, it seems like overrides is a challenge for pnpm even without applying to the full dependency subgraph.  If not (and making the real paths unique to both resolution and override set is doable) then I'm not sure how this override is any different:

```
{
  a: { y: "1.2.3" },
  b: { y: "2.3.4" }
}
```

Reply:
> If so, it seems like overrides is a challenge for pnpm even without applying to the full dependency subgraph. 

Overrides are doable with pnpm if a particular version of a package gets the same set of dependencies. I don't see issues with your last example. Pnpm would create a directory structure like this:

```
.pnpm
+ a@1.0.0
   + node_modules/y -> ../../y@1.2.3
+ b@1.0.0
   + node_modules/y -> ../../y@2.3.4
+ y@2.3.4
+ y@1.2.3
```

But in your first example, we would need to create separate directories for the same x. Something like

```
.pnpm
+ x@1.2.3_<hash1>
   + node_modules/y -> ../../y@1.2.3
+ x@1.2.3_<hash2>
   + node_modules/y -> ../../y@2.3.4
+ y@2.3.4
+ y@1.2.3
```

Note: we actually do this for peer dependencies... but I am not sure it worse adding this for overrides. At the moment, I don't believe overrides need to be so powerful as in this RFC (or as in Yarn).

Reply:
It seems like overrides would be doable in pnpm if a unique combination of (1) resolved version and (2) override set gets the same set of dependencies (which it would).

If you don't feel it's worthwhile to implement the feature in that way, that's really a trade-off between UX, power, and complexity which is really your choice to make.  However, the fact that it *can* be implemented by pnpm makes me feel like this should not be an obstacle to accepting this RFC and npm implementing it with these semantics.